### PR TITLE
Add GUI tools for Clementine

### PR DESCRIPTION
**Pull request by gallottino**
Now it is possible use some kind of graphic debug tools in the UI. 
You can easily add new ones and edit existing ones.

N.B.
I also refactored all the structure of our workspace

### CODE REVIEW

**Comment by guerinoni** [comment]
There are some missing empty line at the end and usually other internale dependencies are not called `packages` but just put in a flat structure like 
```
├── Cargo.lock
├── Cargo.toml
├── src/
│   ├── lib.rs
│   ├── main.rs
│   └── bin/
│       ├── named-executable.rs
│       ├── another-executable.rs
│       └── multi-file-executable/
│           ├── main.rs
│           └── some_module.rs
├── emulator/
│   ├── large-input.rs
│   └── multi-file-bench/
│       ├── main.rs
│       └── bench_module.rs
├── ui/
│   ├── simple.rs
│   └── multi-file-example/
│       ├── main.rs
│       └── ex_module.rs
```

**Comment by guerinoni** [comment]
Another possible useful feature is https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#inheriting-a-dependency-from-a-workspace for managing more than one workspace dependencies

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -29,7 +29,7 @@ jobs:
     steps:
     - uses: actions/checkout@v3
     - name: Test
-      run: cargo test --verbose
+      run: cargo test --verbose --all
```
Comment (guerinoni) [inline .github/workflows/rust.yml:32]:
```
        --workspace                 Test all packages in the workspace
        --all                       Alias for --workspace (deprecated)
```
Better use workspace

File: packages/emu/src/arm7tdmi.rs [hunk 2]
```diff
@@ -102,10 +106,14 @@ impl Cpu for Arm7tdmi {
             self.execute(op_code)
         }
     }
+
+    fn registers(&self) -> Vec<u32> {
```
Comment (guerinoni) [inline packages/emu/src/arm7tdmi.rs:110]: Where is used?

File: .github/workflows/rust.yml (hunk 1 above)
Comment (gallottino) [reply .github/workflows/rust.yml:32]: I'll change it!

File: packages/emu/src/arm7tdmi.rs (hunk 2 above)
Comment (gallottino) [reply packages/emu/src/arm7tdmi.rs:110]: It can be used from some UI widgets: in this case you can see it in cpu_inspector.rs

**Comment by gallottino** [comment]
> There are some missing empty line at the end and usually other internale dependencies are not called `packages` but just put in a flat structure like
> 
> ```
> ├── Cargo.lock
> ├── Cargo.toml
> ├── src/
> │   ├── lib.rs
> │   ├── main.rs
> │   └── bin/
> │       ├── named-executable.rs
> │       ├── another-executable.rs
> │       └── multi-file-executable/
> │           ├── main.rs
> │           └── some_module.rs
> ├── emulator/
> │   ├── large-input.rs
> │   └── multi-file-bench/
> │       ├── main.rs
> │       └── bench_module.rs
> ├── ui/
> │   ├── simple.rs
> │   └── multi-file-example/
> │       ├── main.rs
> │       └── ex_module.rs
> ```

I'm not sure how it works, but my purpose was to divide them into packages mainly for two reasons:
- You can execute test only for emulation or only for ui (and in future for other packages) with: `cargo test -p emu_clementine`
- Emulation it is not only a local dependency but can be also build independently from others. In this way it is possible build custom ui from the emulation package and an user is not forced to use our. 

I don't know what is the best practice, but I wrote this only to clarify my thoughts.

**Comment by guerinoni** [comment]
> > There are some missing empty line at the end and usually other internale dependencies are not called `packages` but just put in a flat structure like
> > ```
> > ├── Cargo.lock
> > ├── Cargo.toml
> > ├── src/
> > │   ├── lib.rs
> > │   ├── main.rs
> > │   └── bin/
> > │       ├── named-executable.rs
> > │       ├── another-executable.rs
> > │       └── multi-file-executable/
> > │           ├── main.rs
> > │           └── some_module.rs
> > ├── emulator/
> > │   ├── large-input.rs
> > │   └── multi-file-bench/
> > │       ├── main.rs
> > │       └── bench_module.rs
> > ├── ui/
> > │   ├── simple.rs
> > │   └── multi-file-example/
> > │       ├── main.rs
> > │       └── ex_module.rs
> > ```
> 
> I'm not sure how it works, but my purpose was to divide them into packages mainly for two reasons:
> 
> * You can execute test only for emulation or only for ui (and in future for other packages) with: `cargo test -p emu_clementine`
> * Emulation it is not only a local dependency but can be also build independently from others. In this way it is possible build custom ui from the emulation package and an user is not forced to use our.
> 
> I don't know what is the best practice, but I wrote this only to clarify my thoughts.

Can I push it behind this?

**Comment by gallottino** [comment]
> > > There are some missing empty line at the end and usually other internale dependencies are not called `packages` but just put in a flat structure like
> > > ```
> > > ├── Cargo.lock
> > > ├── Cargo.toml
> > > ├── src/
> > > │   ├── lib.rs
> > > │   ├── main.rs
> > > │   └── bin/
> > > │       ├── named-executable.rs
> > > │       ├── another-executable.rs
> > > │       └── multi-file-executable/
> > > │           ├── main.rs
> > > │           └── some_module.rs
> > > ├── emulator/
> > > │   ├── large-input.rs
> > > │   └── multi-file-bench/
> > > │       ├── main.rs
> > > │       └── bench_module.rs
> > > ├── ui/
> > > │   ├── simple.rs
> > > │   └── multi-file-example/
> > > │       ├── main.rs
> > > │       └── ex_module.rs
> > > ```
> > 
> > 
> > I'm not sure how it works, but my purpose was to divide them into packages mainly for two reasons:
> > 
> > * You can execute test only for emulation or only for ui (and in future for other packages) with: `cargo test -p emu_clementine`
> > * Emulation it is not only a local dependency but can be also build independently from others. In this way it is possible build custom ui from the emulation package and an user is not forced to use our.
> > 
> > I don't know what is the best practice, but I wrote this only to clarify my thoughts.
> 
> Can I push it behind this?

Of course!

**Review by guerinoni — APPROVED** [review]
Looks good now!

**Pull request merged**